src/persist-client/src/internal/state_versions.rs

```diff
@@ -294,7 +295,7 @@ impl StateVersions {
                         .first()
                         .expect("initialized shard should have at least one diff")
                         .seqno;
-                    all_live_diffs = self.fetch_live_diffs(shard_id).await;
+                    all_live_diffs = self.fetch_recent_live_diffs::<T>(shard_id).await;
```

Comment:
nit: rename this recent_live_diffs please (up in the method sig, github won't let me leave the comment there)

is it crazy to make RecentLiveDiffs and AllLiveDiffs newtype wrappers so we can ensure they're used in the right places? specifically, I just started to go look at the `fetch_current_state` callers to see if they all do the right thing. I think it's at least worth given it a shot and seeing how gross it is or isn't

Reply:
oooh I really really like the newtype approach... the current sticking point is that we sometimes call `fetch_current_state` with compare-and-set's `Err(Vec<VersionedData>)` result that doesn't have a clear contract on whether it's all or recent. hmm...

Reply:
it's a more ambitious change but I wonder if we want `compare_and_set` to no longer own returning the diffs in the mismatch case, then we can consolidate the logic into `StateVersions`

Reply:
wasn't able to find a nice way to slot the newtypes into `fetch_current_state` because some codepaths into that fn go through `fetch_all_live_diffs` and some go through `fetch_recent_live_diffs`. casting `AllLiveDiffs` into `RecentLiveDiffs` when necessary is a possibility, but I didn't love how it looked. 

I updated the latter two fns to use newtypes to add a teeny bit of type safety that at least makes you ponder the meaning of what got returned, but I'm happy to revisit this one further

Reply:
what you have looks great! I have this feeling that StateVersions wants a tiny bit of reworking, but no idea on the specifics of that and certainly not for this PR